lib/signs_ui_web/channels/signs_channel.ex

```diff
@@ -9,42 +9,47 @@ defmodule SignsUiWeb.SignsChannel do
 
   @spec handle_in(String.t(), %{Sign.id() => map()}, any()) :: {:noreply, Phoenix.Socket.t()}
   def handle_in("changeSigns", changes, socket) do
-    if socket_authenticated?(socket) do
-      new_signs = Map.new(changes, fn {id, config} -> {id, Sign.from_config(id, config)} end)
+    case socket_access_level(socket) do
+      :admin ->
+        new_signs = Map.new(changes, fn {id, config} -> {id, Sign.from_config(id, config)} end)
 
-      {:ok, _new_state} = SignsUi.Config.State.update_some(new_signs)
+        {:ok, _new_state} = SignsUi.Config.State.update_some(new_signs)
 
-      username = Guardian.Phoenix.Socket.current_resource(socket)
+        username = Guardian.Phoenix.Socket.current_resource(socket)
 
-      Logger.info("sign_changed: user=#{username} changes=#{inspect(changes)}")
+        Logger.info("sign_changed: user=#{username} changes=#{inspect(changes)}")
 
-      {:noreply, socket}
-    else
-      {:stop, :normal, send_auth_expired_message(socket)}
+        {:noreply, socket}
+
+      :read_only ->
+        {:noreply, socket}
+
+      :none ->
+        {:stop, :normal, send_auth_expired_message(socket)}
     end
   end
 
   intercept(["sign_update"])
 
   def handle_out("sign_update", msg, socket) do
-    if socket_authenticated?(socket) do
+    if socket_access_level(socket) in [:admin, :read_only] do
       push(socket, "sign_update", msg)
       {:noreply, socket}
     else
       {:stop, :normal, send_auth_expired_message(socket)}
     end
   end
 
-  @spec socket_authenticated?(Phoenix.Socket.t()) :: boolean()
-  defp socket_authenticated?(socket) do
+  @spec socket_access_level(Phoenix.Socket.t()) :: SignsUiWeb.AuthManager.access_level()
+  defp socket_access_level(socket) do
     claims = Guardian.Phoenix.Socket.current_claims(socket)
     token = Guardian.Phoenix.Socket.current_token(socket)
 
     with {:ok, claims} <- SignsUiWeb.AuthManager.decode_and_verify(token, claims),
-         true <- SignsUiWeb.AuthManager.claims_grant_signs_access?(claims) do
-      true
+         access_level <- SignsUiWeb.AuthManager.claims_access_level(claims) do
```

Comment:
This probably doesn't need to be in the `with` clauses, since it always matches. You can just remove it from the `with` chain and replace `access_level` inside the `do` block with `SignsUiWeb.AuthManager.claims_access_level(claims)`.

Alternatively, I guess if we always want to ensure that the access level is one of our known levels, you could leave the `with` clause here but add a `when access_level in [...]` guard.

Reply:
Good point. I actually just turned it into a case because I didn't feel like the `with` syntax makes a ton of sense for just one thing.